### PR TITLE
[flang][OpenMP] Convert DataSharingProcessor to omp::Clause

### PR DESCRIPTION
**Pull request by kparzysz**
[Clause representation 6/6]

### CODE REVIEW

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -1384,30 +1386,28 @@ void DataSharingProcessor::copyLastPrivateSymbol(
 }
 
 void DataSharingProcessor::collectOmpObjectListSymbol(
-    const Fortran::parser::OmpObjectList &ompObjectList,
+    const omp::ObjectList &objects,
     llvm::SetVector<const Fortran::semantics::Symbol *> &symbolSet) {
-  for (const Fortran::parser::OmpObject &ompObject : ompObjectList.v) {
-    Fortran::semantics::Symbol *sym = getOmpObjectSymbol(ompObject);
+  for (const omp::Object &object : objects) {
+    Fortran::semantics::Symbol *sym = object.sym;
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP.cpp:1392]: Nit: I think `object.sym` can be passed directly to the call below now that this becomes a trivial statement.

**Review by skatrak — COMMENTED** [review]
LGTM, just small comments.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 2]
```diff
@@ -135,138 +133,135 @@ void DataSharingProcessor::insertBarrier() {
 void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
   bool cmpCreated = false;
   mlir::OpBuilder::InsertPoint localInsPt = firOpBuilder.saveInsertionPoint();
-  for (const Fortran::parser::OmpClause &clause : opClauseList.v) {
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:138]: Are there code changes here or is it only  the replacement of `parser::OmpClause` with `omp::Clause`?

Comment (kparzysz) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:138]: The code functionality is unchanged.  One reason for the seemingly large differences in the diff is that I replaced the `if (get_if<LastPrivate>)` with `if (!get_if<Lastprivate>) continue;` to reduce indentation.

**Review by kiranchandramohan — APPROVED** [review]
LG.

File: flang/lib/Lower/OpenMP.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Done.

File: flang/lib/Lower/OpenMP.cpp [hunk 3]
```diff
@@ -1278,14 +1278,15 @@ class DataSharingProcessor {
   llvm::SetVector<const Fortran::semantics::Symbol *> symbolsInNestedRegions;
   llvm::SetVector<const Fortran::semantics::Symbol *> symbolsInParentRegions;
   Fortran::lower::AbstractConverter &converter;
+  Fortran::semantics::SemanticsContext &semaCtx;
```
[the remark this answers is not in the document]
Comment (kparzysz) [reply flang/lib/Lower/OpenMP.cpp]: Removed.

**Review by tblah — APPROVED** [review]

**Pull request merged**